File: ch02/src/cpp/relaxed_ordering_progress_reporting_multiworkers.cpp
#include <atomic>
#include <chrono>
#include <cstdint>
#include <iostream>
#include <thread>
#include <vector>

using namespace std::chrono_literals;

void process_item(int)
{
    std::this_thread::sleep_for(50ms);
}

int main()
{
    auto num_done = std::atomic<std::uint64_t> { 0 };
    std::vector<std::jthread> background_threads;
    for (auto t { 0 }; t < 4; ++t) {
        background_threads.emplace_back([&]() {
            for (auto i { 0 }; i < 25; ++i) {
                process_item(25 * t + i);
                num_done.fetch_add(1, std::memory_order_relaxed);
            }
        });
    }

    while (true) {
        const auto n = num_done.load(std::memory_order_relaxed);
        std::cout << "Working ... " << n << "/100 done\n";
        if (n == 100) {
            break;
        }

        std::this_thread::sleep_for(500ms);
    }

    return 0;
}

File: ch02/src/bin/relaxed_ordering_progress_reporting_multiworkers.rs
use std::{
    sync::atomic::{AtomicUsize, Ordering},
    thread,
    time::{Duration, Instant},
};

/// Example or relaxed memory ordering.
/// This order is OK because the two threads operate on the same
/// variable.
fn main() {
    let num_done = &AtomicUsize::new(0);

    thread::scope(|s| {
        // 4 background threads doing the work
        for t in 0..4 {
            s.spawn(move || {
                for i in 0..25 {
                    process_item(t * 25 + i);
                    num_done.fetch_add(1, Ordering::Relaxed);
                }
            });
        }

        let start = Instant::now();
        // The main thread shows status update
        loop {
            let n = num_done.load(Ordering::Relaxed);
            println!(
                "{:.2?} Working... {n}/100 done",
                Instant::now().duration_since(start)
            );
            if n == 100 {
                break;
            }

            thread::sleep(Duration::from_secs(1));
        }
    })
}

fn process_item(_: usize) {
    thread::sleep(Duration::from_millis(400));
}
